File: rs2/src/main/java/SoundPlayer.java
import javax.sound.sampled.*;
import java.io.InputStream;
import java.util.HashMap;


public class SoundPlayer {

	public static HashMap<Integer, AudioInputStream> sounds = new HashMap<>();

	private AudioInputStream stream;
	private DataLine.Info info;
	private Clip sound;

	private InputStream soundStream;
	private int delay;
	private int soundLevel;
	private InputStream arg0;
	public static int volume;

	public SoundPlayer(AudioInputStream stream, int level, int delay) {
		if (level == 0 || volume == 4 || level - volume <= 0) {
			return;
		}
		this.soundStream = stream;
		this.soundLevel = level;
		this.delay = delay;
		run();
	}

	public void run() {
		try {
			stream = (AudioInputStream) soundStream;
			info = new DataLine.Info(Clip.class, stream.getFormat());
			sound = (Clip) AudioSystem.getLine(info);
			sound.open(stream);
			FloatControl gain = (FloatControl) sound.getControl(FloatControl.Type.MASTER_GAIN);
			gain.setValue(getDecibels(gain, signlink.wavevol));
			if (delay > 0) {
				Thread.sleep(delay);
			}
			sound.start();
			sound.drain();
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	public float getDecibels(FloatControl volumeControl, int level) {
		float range = volumeControl.getMaximum() - volumeControl.getMinimum();
		float min = volumeControl.getMinimum();
		switch (level) {
			case 0: // 4 in player options
				return min + (range * .90f);
			case -400: // 3
				return min + (range * .74f);
			case -800: // 2
				return min + (range * .62f);
			case -1200: // 1
				return min + (range * .50f);
			default:
				return min;
		}
	}
}

File: composeApp/src/desktopMain/kotlin/meteor/audio/SoundPlayer.kt
package meteor.audio

import meteor.Main
import meteor.events.PlaySound
import org.rationalityfrontline.kevent.KEVENT
import java.io.ByteArrayInputStream
import java.io.InputStream
import javax.sound.sampled.*

class SoundPlayer(stream: AudioInputStream, delay: Int) {
    private var stream: AudioInputStream? = null
    private var info: DataLine.Info? = null
    private var sound: Clip? = null

    private var soundStream: InputStream? = null
    private var delay: Int? = null

    init {
        this.soundStream = stream
        this.delay = delay
        run()
    }

    fun run() {
        try {
            stream = soundStream as AudioInputStream
            info = DataLine.Info(Clip::class.java, stream!!.format)
            sound = AudioSystem.getLine(info) as Clip
            sound!!.open(stream)
            val gain = sound!!.getControl(FloatControl.Type.MASTER_GAIN) as FloatControl
            gain.value = getDecibels(gain, Main.client.waveVol)
            if (delay!! > 0) {
                Thread.sleep(delay!!.toLong())
            }
            sound!!.start()
            sound!!.drain()
            sound!!.flush()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun getDecibels(volumeControl: FloatControl, level: Int): Float {
        val range = volumeControl.maximum - volumeControl.minimum
        val min = volumeControl.minimum
        return when (level) {
            0 -> min + (range * .90f)
            -400 -> min + (range * .74f)
            -800 -> min + (range * .62f)
            -1200 -> min + (range * .50f)
            else -> min
        }
    }
}